apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -102,7 +112,70 @@ type ComposedTemplate struct {
 	ReadinessChecks []ReadinessCheck `json:"readinessChecks,omitempty"`
 }
 
-// TypeReadinessCheck is used for readiness check types
+// mapPatchSets processes a list of PatchSets and returns a map of PatchSet
+// ID to Patches.
+func mapPatchSets(patchSets []PatchSet) (patchSetMap, error) {
+	var m = patchSetMap{}
+	for _, ps := range patchSets {
+		// Make sure each patch inside patch set is valid
+		// PatchSets cannot contain reference patches (how would they be resolved?)
+		for i := range ps.Patches {
+			ps.Patches[i].SetDefaults()
+			if ps.Patches[i].Type == PatchTypeReference {
+				return nil, errors.Errorf(errInvalidPatchSetPatch, i, ps.ID)
+			}
+		}
+		// Map patches against patch set ID
+		m[ps.ID] = ps.Patches
+	}
+	return m, nil
+}
+
+// InjectCompositionData performs processing tasks on a composed template
+// prior to rendering. It is is executed as part of the reconciliation
+// loop, just after the ComposedTemplate is instantiated, and is passed
+// the Composition Spec - this allows data to be passed between the composition
+// and the composed, such as PatchSets.
+func (c *ComposedTemplate) InjectCompositionData(comp CompositionSpec) error {
```

Comment:
`InjectCompositionData` feels like we're bringing in a blob of data to be used in spec of the managed resources but in reality we just inject patches. Maybe `DereferencePatches` or `FlattenPatches`?

Reply:
I had a similar thought, but preferred to stick with `InjectCompositionData` since that name would make sense if this method ever did more than inject patch sets.

Reply:
Yeah my original reasoning behind `InjectCompositionData` was that it applied a new integration point between the composition and the composed resources, that _at the moment_ only does `PatchSet` injection.

Thinking about this more, I wonder if in future there would ever be movement in the reverse direction, in which case `InjectCompositionData` might also be too specific in a single direction.

Reply:
I'm happy to defer to your judgement on naming here @benagricola. I don't think we need anything too future proof - we can always rename it if its function changes in future. :)